lib/dry/types/coercions.rb

```diff
@@ -8,18 +8,21 @@ def to_nil(input)
       end
 
       def to_date(input)
+        return input unless input.respond_to?(:to_str)
         Date.parse(input)
       rescue ArgumentError
         input
       end
 
       def to_date_time(input)
+        return input unless input.respond_to?(:to_str)
         DateTime.parse(input)
       rescue ArgumentError
         input
       end
 
       def to_time(input)
+        return input unless input.respond_to?(:to_str)
         Time.parse(input)
       rescue ArgumentError
         input
```

Comment:
Please just add `TypeError` to the rescue clauses (that's how it's done for other types like `Int` etc.). This should cover all cases where it may fail.